Make cluster controller node state changes by default wait until change is visible in cluster

vekterli opened the pull request:
@hakonhall please review. I've tried to generalize the waiting criteria a bit more since our discussion.

Note that this PR does not cover the `vespa-set-node-state` client tool. We want the default behavior to be "wait until ACKed", which is what it will get until any client override support is in place.

It also does not cover the desired change in logic where we do not require all distributors to have ACKed the exact same version, but rather look at the minimum ACKed version across all distributors. This is more of an availability optimization, and is not required for correctness.

vekterli committed:
Add support for version ACK-dependent tasks in cluster controller

Used to enable synchronous operation for set-node-state calls, which
ensure that side-effects of the call are visible when the response
returns.

If controller leadership is lost before state is published, tasks will
be failed back to the client.

vekterli committed:
Test automatic task failing on controller leadership loss

vekterli committed:
Move leadership test code into fixture

vekterli committed:
Test multiple scheduled synchronous tasks

hakonhall commented on clustercontroller-utils/src/main/java/com/yahoo/vespa/clustercontroller/utils/staterestapi/requests/SetUnitStateRequest.java:
```diff
@@ -16,17 +16,42 @@
     Condition getCondition();
+
+    enum ResponseWait {
+        WAIT_UNTIL_CLUSTER_ACKED("wait-until-cluster-acked"), // Wait for state change to be ACKed by cluster
```
Should also indicate this is implicit if omitted.

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/FleetController.java:
```diff
@@ -727,10 +770,31 @@ private boolean recomputeClusterStateIfRequired() {
+     * This works transparently for tasks that end up changing the current cluster state (i.e.
+     * requiring a new state to be published) and for those whose changes are idempotent. In the
+     * former case the tasks will depend on the version that was generated based upon them. In
+     * the latter case the tasks will depend on the version that is already published (or in the
```
Agreed. In this context perhaps calling it a "no-op" is more semantically accurate.

vekterli committed:
Break node version ACK check out into separately called logic

Removes dependency on having to invoke broadcastNewState before being
able to observe that all distributors are in sync. Invocations of
broadcastNewState are gated by a grace period between each time,
so unless this is done we get artificial delays before a synchronous
task can be considered complete.

vekterli committed:
Add comment that cluster state ACK waiting is the default

vekterli committed:
Change wording for operations without observable side-effects

vekterli committed:
Refactor deferred version task completion to take in version explicitly

vekterli commented:
@hakonhall PTAL
